### PR TITLE
Image resizing

### PR DESCRIPTION
**Pull request by smauq**
Added flags to enable resizing the input images for rovioli and maplab node. The image resize is carried over to the saved sensor config when map building.

Depends on this tiny PR in aslam_cv [#54](https://github.com/ethz-asl/aslam_cv2/pull/54)

### CODE REVIEW

File: applications/maplab-node/src/maplab-node.cc [hunk 1]
```diff
@@ -43,6 +43,8 @@ DEFINE_int64(
     "Enable visualization of the visual localization observations and inliers "
     "as ROS message.");
 
+DECLARE_double(image_resize_factor);
```
Comment (floriantschopp) [inline applications/maplab-node/src/maplab-node.cc:46]: Should we give it a more clear name? like for rovioli and maplab independently?

File: applications/maplab-node/src/maplab-node.cc [hunk 2]
```diff
@@ -65,6 +67,31 @@ MaplabNode::MaplabNode(
   }
   CHECK(sensor_manager_);
 
+  // At the very beginning check if the main ncamera images need to be resized
```
Comment (floriantschopp) [inline applications/maplab-node/src/maplab-node.cc:70]: .

File: applications/maplab-node/src/maplab-node.cc [hunk 3]
```diff
@@ -65,6 +67,31 @@ MaplabNode::MaplabNode(
...
+  aslam::NCamera::Ptr mapping_ncamera =
+      vi_map::getSelectedNCamera(*sensor_manager_);
+  if (mapping_ncamera && fabs(FLAGS_image_resize_factor - 1.0) > 1e-6) {
+    for (size_t i = 0; i < mapping_ncamera->getNumCameras(); i++) {
+      // The intrinsics of the camera can just be multiplied with the resize
+      // factor. Distortion parameters are agnostic to the image size
```
Comment (floriantschopp) [inline applications/maplab-node/src/maplab-node.cc:76]: .

File: applications/maplab-node/src/maplab-node.cc [hunk 4]
```diff
@@ -65,6 +67,31 @@ MaplabNode::MaplabNode(
...
+          camera->getParameters() * FLAGS_image_resize_factor);
+      camera->setImageWidth(
+          round(camera->imageWidth() * FLAGS_image_resize_factor));
+      camera->setImageHeight(
+          round(camera->imageHeight() * FLAGS_image_resize_factor));
+      camera->setDescription(
+          camera->getDescription() + " - resized " +
+          std::to_string(FLAGS_image_resize_factor));
+
+      // The parameters have changed so we need to generate a new sensor id
```
Comment (floriantschopp) [inline applications/maplab-node/src/maplab-node.cc:88]: .

File: applications/maplab-node/src/ros-helpers.cc [hunk 5]
```diff
@@ -112,6 +114,17 @@ vio::ImageMeasurement::Ptr convertRosImageToMaplabImage(
   }
   CHECK(cv_ptr);
 
+  if (fabs(FLAGS_image_resize_factor - 1.0) > 1e-6) {
+    int newcols =
```
Comment (floriantschopp) [inline applications/maplab-node/src/ros-helpers.cc:118]: `static const`?

File: applications/maplab-node/src/ros-helpers.cc [hunk 6]
```diff
@@ -112,6 +114,17 @@ vio::ImageMeasurement::Ptr convertRosImageToMaplabImage(
...
+        round(image_measurement->image.cols * FLAGS_image_resize_factor);
+    int newrows =
```
Comment (floriantschopp) [inline applications/maplab-node/src/ros-helpers.cc:120]: `static const`?

File: applications/maplab-node/src/maplab-node.cc (hunk 1 above)
Comment (floriantschopp) [reply applications/maplab-node/src/maplab-node.cc:46]: as you have `rovioli_image_resize_factor` later, rename this to `maplab_image_resize_factor`

File: applications/rovioli/app/rovioli-app.cc [hunk 7]
```diff
@@ -65,6 +67,29 @@ int main(int argc, char** argv) {
   CHECK(vi_map::getSelectedImu(sensor_manager))
       << "[ROVIOLI] The sensor calibration does not contain an IMU!";
 
+  if (fabs(FLAGS_rovioli_image_resize_factor - 1.0) > 1e-6) {
+    aslam::NCamera::Ptr ncamera = vi_map::getSelectedNCamera(sensor_manager);
```
Comment (floriantschopp) [inline applications/rovioli/app/rovioli-app.cc:71]: can we have the same convention here as in maplab_node? with `mapping_ncamera` before if and check if its valid?

File: applications/rovioli/app/rovioli-app.cc [hunk 8]
```diff
@@ -65,6 +67,29 @@ int main(int argc, char** argv) {
...
+    for (size_t i = 0; i < ncamera->getNumCameras(); i++) {
+      // The intrinsics of the camera can just be multiplied with the resize
+      // factor. Distortion parameters are agnostic to the image size
```
Comment (floriantschopp) [inline applications/rovioli/app/rovioli-app.cc:74]: .

File: applications/rovioli/app/rovioli-app.cc [hunk 9]
```diff
@@ -65,6 +67,29 @@ int main(int argc, char** argv) {
...
+      aslam::Camera::Ptr camera = ncamera->getCameraShared(i);
+      camera->setParameters(
+          camera->getParameters() * FLAGS_rovioli_image_resize_factor);
+      camera->setImageWidth(
+          round(camera->imageWidth() * FLAGS_rovioli_image_resize_factor));
+      camera->setImageHeight(
+          round(camera->imageHeight() * FLAGS_rovioli_image_resize_factor));
+      camera->setDescription(
+          camera->getDescription() + " - resized " +
```
Comment (floriantschopp) [inline applications/rovioli/app/rovioli-app.cc:83]: +1 :-)

File: applications/rovioli/app/rovioli-app.cc [hunk 10]
```diff
@@ -65,6 +67,29 @@ int main(int argc, char** argv) {
...
+          std::to_string(FLAGS_rovioli_image_resize_factor));
+
+      // The parameters have changed so we need to generate a new sensor id
```
Comment (floriantschopp) [inline applications/rovioli/app/rovioli-app.cc:86]: .

File: applications/rovioli/src/ros-helpers.cc [hunk 11]
```diff
@@ -108,6 +110,17 @@ vio::ImageMeasurement::Ptr convertRosImageToMaplabImage(
   }
   CHECK(cv_ptr);
 
+  if (fabs(FLAGS_rovioli_image_resize_factor - 1.0) > 1e-6) {
+    int newcols = round(
```
Comment (floriantschopp) [inline applications/rovioli/src/ros-helpers.cc:114]: `static const`?

File: applications/rovioli/src/ros-helpers.cc [hunk 12]
```diff
@@ -108,6 +110,17 @@ vio::ImageMeasurement::Ptr convertRosImageToMaplabImage(
...
+        image_measurement->image.cols * FLAGS_rovioli_image_resize_factor);
+    int newrows = round(
```
Comment (floriantschopp) [inline applications/rovioli/src/ros-helpers.cc:116]: `static const`?

File: applications/maplab-node/src/maplab-node.cc (hunk 1 above)
Comment (smauq) [reply applications/maplab-node/src/maplab-node.cc:46]: Sadly this seems to be our convention at the moment if you look around at the other flags...

File: applications/maplab-node/src/maplab-node.cc [hunk 13]
```diff
@@ -65,6 +67,31 @@ MaplabNode::MaplabNode(
   }
   CHECK(sensor_manager_);
 
+  // At the very beginning check if the main ncamera images need to be resized.
+  aslam::NCamera::Ptr mapping_ncamera =
+      vi_map::getSelectedNCamera(*sensor_manager_);
+  if (mapping_ncamera && fabs(FLAGS_image_resize_factor - 1.0) > 1e-6) {
```
Comment (floriantschopp) [inline applications/maplab-node/src/maplab-node.cc:73]: Tiny: maybe also convert the `if` of the `mapping_camera` into a check as below.

Comment (smauq) [reply applications/maplab-node/src/maplab-node.cc:73]: The camera is optional here if you check later when the visual mapping is initialized

**Pull request merged**